dateparser/utils.py

```diff
@@ -86,3 +87,36 @@ def find_date_separator(format):
     m = re.search(r'(?:(?:%[dbBmaA])(\W))+', format)
     if m:
         return m.group(1)
+
+
+def registry(cls):
+    def choose(creator):
+        def constructor(cls, *args, **kwargs):
+            key = cls.get_key(*args, **kwargs)
+
+            if not hasattr(cls, "__registry_dict"):
+                setattr(cls, "__registry_dict", {})
+            registry_dict = getattr(cls, "__registry_dict")
+
+            if key not in registry_dict:
+                registry_dict[key] = creator(cls, *args, **kwargs)
```

Comment:
Consider having `setattr(registry_dict[key], 'registry_key', key)` here, in case you need this information for some refactoring.